Some cleanup and new features.

nyscotsman opened the pull request:
Cleanup - parse out attributes when constructing the graph; use the graph object itself to generate the .dot file.

Features - added some decorators (late, active, up next) to task nodes. Base formatting on task status.

A "magic" Start task and support for dates offset from the project start makes it easier to show these new features with the test data.

nyscotsman committed:
Cleanup - parse out attributes when constructing the graph; use the graph to generate the .dot file.
Features - added some decorators (late, active, up next) to task nodes. Base formatting on task status.

nyscotsman commented:
![image](https://github.com/user-attachments/assets/aa5bca2d-b038-48b4-9baf-e599d7c4d8e3)

Done tasks just show their end date; milestones just their start date. (estimate should be zero on milestones; I was thinking perhaps the "milestone" status should be the criteria, and we can calculate an actual milestone status based on the status of their deps?).

Active, late and up next tasks have an annotation on their title. Moved status up between the start and end dates for in progress tasks. Added a # bus days available for a task to the estimate box (so you can compare work to available time).

johnlevidy commented on backend/graph.py:
```diff
@@ -74,38 +134,44 @@ def find_start_next_before_end(G: nx.Graph, notifications):
-      total_length, critical_path_length = compute_dag_metrics(G)
-      parallelism_ratio = total_length / critical_path_length
-      notifications.append(Notification(Severity.INFO, f"[Total Length: {total_length}], [Critical Path Length: {critical_path_length}], [Parallelism Ratio: {parallelism_ratio:.2f}]"))
+        raise Exception("Can't build graph.")
```
Did you test this case on the front end? Sometimes Exceptions behave weirdly. Mostly, I want to make sure the notifications still display if you raise here.

nyscotsman replied:
It's called from app.py within a try/except block. I think actually there's a bug there too; looks like verify_schema is supposed to throw on failure but it actually returns a status bool. I can fix that later.

johnlevidy commented on backend/dot.py:
woah look at you w/ the new features.

nyscotsman replied:
😄

johnlevidy commented on backend/dot.py:
at some point we should class-ify task and pass the class to this function ( in node['value'] or something )

nyscotsman replied:
Agreed; for now I switched to a StrEnum to tidy it up a bit (see the new commit).

johnlevidy commented on backend/graph.py:
```diff
@@ -4,35 +4,95 @@
     to_return = False
     for task_name, task in G.nodes(data=True):
-        if not task['StartDate'] or not task['EndDate']:
+        if not task['start_date'] or not task['end_date']:
```
just a note, this may require spreadsheet template changes.

johnlevidy replied:
just checked this code out -- actually this causes several tests to break as well. we can pull it out into a constant if we think this might change again.

nyscotsman replied:
It's fine for the spreadsheet because these keys are for the graph nodes, not the input data. I didn't run the tests 😢 , let me look at those.

nyscotsman replied:
all tests are passing...? (maybe my next commit fixed something)

johnlevidy reviewed (changes requested):
Overall new features and refactors look great.

Two main asks:

1. make sure front end works in the cycle-detected case. ( mostly that notifications still display )
2. run / fix the pytests, I think it's just a matter of swapping snake + camel.

nyscotsman committed:
Add a StrEnum for the Graph node attributes we're using.
Populate any missing dates and/or estimates.
Render estimated values in italics in the svg.

nyscotsman commented:
![image](https://github.com/user-attachments/assets/4e8d623d-702c-43fe-b05e-8f3ed10f2dd4)

Shows generated dates in italics.
